File: cpp/cpu_benchmark.cpp
#include "cpu_benchmark.h"
#include <iostream>
#include <cmath>
#include <random>
#include <chrono>
#include <future>
#include <vector>

// ============================================
// ① 乱数の初期化をスレッドローカルで行う（高速化）
// ============================================
double heavyComputation(int n) {
    static thread_local std::mt19937 gen(std::random_device{}()); // 各スレッド専用の RNG
    std::uniform_real_distribution<> dis(0.0, 1.0);

    double sum = 0.0;
    for (int i = 0; i < n; ++i) {
        sum += std::sqrt(i * dis(gen));
    }
    return sum;
}

// ============================================
// ② 並列化（std::async によるマルチスレッド実行）
// ============================================
void runCPUBenchmark(int taskCount) {
    std::cout << "C++ benchmark start" << std::endl;

    auto start = std::chrono::high_resolution_clock::now();
    int iterationsPerTask = 10'000'000; // 1千万回

    std::vector<std::future<double>> futures;
    futures.reserve(taskCount);

    // 並列タスクを起動
    for (int i = 0; i < taskCount; ++i) {
        futures.push_back(std::async(std::launch::async, [iterationsPerTask]() {
            return heavyComputation(iterationsPerTask);
        }));
    }

    // 全タスクの完了を待機
    for (int i = 0; i < taskCount; ++i) {
        double result = futures[i].get();
        std::cout << "Task " << (i + 1) << " done" << std::endl;
    }

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::duration<double>>(end - start);
    std::cout << "All " << taskCount << " tasks done in " << duration.count() << "s" << std::endl;
}

File: rust/src/cpu_benchmark.rs
use std::time::Instant;
use rand::{Rng, SeedableRng};
use rand::rngs::SmallRng;

fn heavy_computation(n: usize) -> f64 {
    let mut sum = 0.0;
    // let mut rng = rand::thread_rng();
    // → セキュリティに考慮したスレッドセーフで安全な乱数生成器だが、コストが高いため重くなるそう
    // 代わりに ↓ を使うとパフォーマンスが劇的に向上した
    let mut rng = SmallRng::from_entropy();
    for i in 0..n {
        sum += (i as f64 * rng.gen::<f64>()).sqrt();
    }
    sum
}

pub fn run_cpu_benchmark(task_count: usize) {
    println!("Rust benchmark start (optimized)");

    let start = Instant::now();
    let iterations_per_task = 10_000_000;

    for i in 0..task_count {
        let _result = heavy_computation(iterations_per_task);
        println!("Task {} done", i + 1);
    }

    let duration = start.elapsed();
    println!("All {} tasks done in {:?}", task_count, duration);
}